app/src/main/java/org/mozilla/fenix/library/history/HistoryDataSource.kt

```diff
@@ -5,38 +5,36 @@
 package org.mozilla.fenix.library.history
 
 import androidx.annotation.VisibleForTesting
-import androidx.paging.ItemKeyedDataSource
+import androidx.paging.PagingSource
+import androidx.paging.PagingState
 import org.mozilla.fenix.components.history.HistoryDB
 import org.mozilla.fenix.components.history.PagedHistoryProvider
 
 class HistoryDataSource(
     private val historyProvider: PagedHistoryProvider
-) : ItemKeyedDataSource<Int, History>() {
+) : PagingSource<Int, History>() {
 
-    // Because the pagination is not based off of the key
-    // we want to start at 1, not 0 to be able to send the correct offset
-    // to the `historyProvider.getHistory` call.
-    override fun getKey(item: History): Int = item.position + 1
+    // having any value but null creates visual glitches in case or swipe to refresh and immediate
+    // scroll down
+    override fun getRefreshKey(state: PagingState<Int, History>): Int? = null
 
-    override fun loadInitial(
-        params: LoadInitialParams<Int>,
-        callback: LoadInitialCallback<History>
-    ) {
-        historyProvider.getHistory(INITIAL_OFFSET, params.requestedLoadSize) { history ->
-            callback.onResult(history.positionWithOffset(INITIAL_OFFSET))
+    // params.key is expected to be null for the initial load or a refresh
+    override suspend fun load(params: LoadParams<Int>): LoadResult<Int, History> {
+        val offset = params.key ?: 0
```

Comment:
I am wondering if it should be 1 instead of 0 from looking at the sample code in https://developer.android.com/topic/libraries/architecture/paging/v3-paged-data#pagingsource

Reply:
In the example data source supports pagination: `backend.searchUsers(query, nextPageNumber)`
, but in our case – `historyProvider.getHistory` operates using an offset. 
So the code actually operates with the same concept. That goes like 0/75/100/125 (in case there are no duplicates)
```
return LoadResult.Page(
            data = historyItems,
            prevKey = null, // Only paging forward.
            nextKey = nextOffset
        )
```